File: GCDLib/UI/FISLibrary/frmAddFIS.vb
Namespace UI.FISLibrary

    Public Class frmAddFIS

        Private Sub AddFISForm_Load(sender As Object, e As System.EventArgs) Handles Me.Load
            FISTableBindingSource.DataSource = Core.GCDProject.ProjectManager.fisds
            FISTableBindingSource.AddNew()
        End Sub

        Private Sub btnOK_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnOK.Click

            Try
                If txtName.TextLength < 1 Then
                    MsgBox("Please enter a name for the FIS file.", MsgBoxStyle.Exclamation, My.Resources.ApplicationNameLong)
                    Me.DialogResult = System.Windows.Forms.DialogResult.None
                    Exit Sub
                End If

                If txtFISFile.TextLength < 1 Then
                    MsgBox("Please select a FIS file.", MsgBoxStyle.Exclamation, My.Resources.ApplicationNameLong)
                    Me.DialogResult = System.Windows.Forms.DialogResult.None
                    Exit Sub
                End If

                If Not System.IO.File.Exists(txtFISFile.Text) Then
                    System.Windows.Forms.MessageBox.Show("The FIS file does not exist.", My.Resources.ApplicationNameLong, System.Windows.Forms.MessageBoxButtons.OK, System.Windows.Forms.MessageBoxIcon.Information)
                    Me.DialogResult = System.Windows.Forms.DialogResult.None
                    Exit Sub
                End If

                Try
                    Dim theFile As New Core.ErrorCalculation.FIS.FISRuleFile(txtFISFile.Text)

                Catch ex As Exception
                    System.Windows.Forms.MessageBox.Show("The FIS file is invalid and/or badly formatted. Check that the formatting of the file contents match the MatLab fully inference toolbox specifications and try again.", My.Resources.ApplicationNameLong, System.Windows.Forms.MessageBoxButtons.OK, System.Windows.Forms.MessageBoxIcon.Information)
                    Me.DialogResult = System.Windows.Forms.DialogResult.None
                    Exit Sub
                End Try

                Dim CurrentRow As DataRowView = FISTableBindingSource.Current

                CurrentRow("Name") = txtName.Text
                CurrentRow("Path") = txtFISFile.Text

                FISTableBindingSource.EndEdit()
                Core.GCDProject.ProjectManager.saveFIS()

            Catch ex As Exception
                If ex.Message.ToString.ToLower.Contains("name") Then
                    MsgBox("Please select a unique name for the selected FIS file.", MsgBoxStyle.Exclamation, My.Resources.ApplicationNameLong)
                ElseIf ex.Message.ToString.ToLower.Contains("path") Then
                    MsgBox("A FIS file with the same name is already present in the FIS library.", MsgBoxStyle.Exclamation, My.Resources.ApplicationNameLong)
                Else
                    MsgBox("An error occured while trying to save the information, " & vbNewLine & ex.Message)
                End If
                DialogResult = System.Windows.Forms.DialogResult.None
            End Try

        End Sub

        Private Sub btnBrowseFIS_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnBrowseFIS.Click

            Dim fileDialog As System.Windows.Forms.OpenFileDialog = New System.Windows.Forms.OpenFileDialog()
            fileDialog.Title = "Select a FIS File"
            fileDialog.Filter = "GCD FIS Files (*.fis) | *.fis"
            fileDialog.InitialDirectory =
        fileDialog.RestoreDirectory = False

            If fileDialog.ShowDialog() = DialogResult.OK Then
                txtFISFile.Text = fileDialog.FileName

                If String.IsNullOrEmpty(txtName.Text) Then
                    txtName.Text = IO.Path.GetFileNameWithoutExtension(fileDialog.FileName)
                    txtName.SelectAll()
                    txtName.Focus()
                End If
            End If

        End Sub

        Private Sub btnHelp_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnHelp.Click
            Process.Start(My.Resources.HelpBaseURL & "gcd-command-reference/customize-menu/fis-library")
        End Sub
    End Class

End Namespace

File: GCDCore/UserInterface/FISLibrary/frmAddFIS.cs
using System.Windows.Forms;
using System;
using System.Diagnostics;
using GCDCore.ErrorCalculation.FIS;
using System.Linq;

namespace GCDCore.UserInterface.FISLibrary
{
    public partial class frmAddFIS
    {
        public FISLibraryItem FISItem { get; internal set; }

        public frmAddFIS()
        {
            InitializeComponent();
        }


        private void AddFISForm_Load(object sender, System.EventArgs e)
        {
        }

        private void btnOK_Click(System.Object sender, System.EventArgs e)
        {
            if (!ValidateForm())
            {
                this.DialogResult = System.Windows.Forms.DialogResult.None;
                return;
            }

            try
            {
                FISItem = new FISLibraryItem(txtName.Text, new System.IO.FileInfo(txtFISFile.Text));
            }
            catch (Exception ex)
            {
                if (ex.Message.ToString().ToLower().Contains("name"))
                {
                    MessageBox.Show("Please select a unique name for the selected FIS file.", Properties.Resources.ApplicationNameLong, MessageBoxButtons.OK, MessageBoxIcon.Exclamation);
                }
                else if (ex.Message.ToString().ToLower().Contains("path"))
                {
                    MessageBox.Show("A FIS file with the same name is already present in the FIS library.", Properties.Resources.ApplicationNameLong, MessageBoxButtons.OK, MessageBoxIcon.Exclamation);
                }
                else
                {
                    MessageBox.Show("An error occured while trying to save the information, " + Environment.NewLine + ex.Message);
                }
                DialogResult = System.Windows.Forms.DialogResult.None;
            }
        }

        private bool ValidateForm()
        {
            // Santiy check to avoid empty string names
            txtName.Text = txtName.Text.Trim();

            if (txtName.TextLength < 1)
            {
                MessageBox.Show("Please enter a name for the FIS file.", Properties.Resources.ApplicationNameLong, MessageBoxButtons.OK, MessageBoxIcon.Exclamation);
                return false;
            }

            if (txtFISFile.TextLength < 1)
            {
                MessageBox.Show("Please select a FIS file.", Properties.Resources.ApplicationNameLong, MessageBoxButtons.OK, MessageBoxIcon.Exclamation);
                return false;
            }

            if (!System.IO.File.Exists(txtFISFile.Text))
            {
                MessageBox.Show("The FIS file does not exist.", Properties.Resources.ApplicationNameLong, MessageBoxButtons.OK, MessageBoxIcon.Information);
                return false;
            }

            if (GCDCore.Project.ProjectManager.FISLibrary.Count<FISLibraryItem>(x => string.Compare(x.Name, txtName.Text,true)==0) > 0)
            {
                MessageBox.Show("A FIS library item already exists with this name. Each FIS library item must have a unique name.", Properties.Resources.ApplicationNameLong, MessageBoxButtons.OK, MessageBoxIcon.Information);
                txtName.Select();
                return false;
            }
            
            if (GCDCore.Project.ProjectManager.FISLibrary.Count<FISLibraryItem>(x => string.Compare(txtFISFile.Text, x.FilePathString,true)==0) > 0)
            {
                MessageBox.Show("A FIS library item already refers to this FIS rule file. Each FIS library item must refere to a different file.", Properties.Resources.ApplicationNameLong, MessageBoxButtons.OK, MessageBoxIcon.Information);
                return false;
            }

            try
            {
                ErrorCalculation.FIS.FISRuleFile theFile = new ErrorCalculation.FIS.FISRuleFile(new System.IO.FileInfo(txtFISFile.Text));
            }
            catch (Exception ex)
            {
                System.Windows.Forms.MessageBox.Show("The FIS file is invalid and/or badly formatted. Check that the formatting of the file contents match the MatLab fully inference toolbox specifications and try again.", GCDCore.Properties.Resources.ApplicationNameLong, System.Windows.Forms.MessageBoxButtons.OK, System.Windows.Forms.MessageBoxIcon.Information);
                return false;
            }

            return true;
        }

        private void btnBrowseFIS_Click(System.Object sender, System.EventArgs e)
        {
            System.Windows.Forms.OpenFileDialog fileDialog = new System.Windows.Forms.OpenFileDialog();
            fileDialog.Title = "Select a FIS File";
            fileDialog.Filter = "GCD FIS Files (*.fis) | *.fis";
            fileDialog.RestoreDirectory = false;

            if (fileDialog.ShowDialog() == DialogResult.OK)
            {
                txtFISFile.Text = fileDialog.FileName;

                if (string.IsNullOrEmpty(txtName.Text))
                {
                    txtName.Text = System.IO.Path.GetFileNameWithoutExtension(fileDialog.FileName);
                    txtName.SelectAll();
                    txtName.Focus();
                }
            }
        }

        private void btnHelp_Click(System.Object sender, System.EventArgs e)
        {
            Process.Start(Properties.Resources.HelpBaseURL + "gcd-command-reference/customize-menu/fis-library");
        }
    }
}
